### ISSUE TITLE
Lacking proper guidance for customers who used to inject IBus in their custom classes when using NServiceBus.Host

### ISSUE THREAD

**Issue by indualagarsamy** [Particular/docs.particular.net#1497]
Relates to https://github.com/Particular/V6Launch/issues/27

**IBus is gone. What now?**
From a NServiceBus.Host perspective, for people that used to get IBus injected into their own containers,what now?

Should they be using `IEndpointInstance`?  If they are self hosting, they can get the `IEndpointInstance` when they create the bus and the documentation is clear, however how does one get to the `IEndpointInstance` when using the NServiceBus.Host? When using the NServiceBus.Host's `IWantToRunWhenEndpointStartsAndStops`, there's the `IManageSession` interface. Should the users use that in lieu of what was `IBus`? 

There's no documentation on `IManageSession`. Nothing in the upgrade guides either. 
So, from a user perspective, who only knew `IBus` before, now there's `IEndpointInstance`, `IManageSession` and the `IHandlerContext`. What should they use if they want to hang on the concept of IBus like before? 

- Also previously IBus was injected:

```c#
[Inject]
public IBus Bus {get; set;}
```

Does it now become?

```c#
[Inject]
public IEndpointInstance EndpointInstance {get; set;}
```

or

```c#
[Inject]
public IManageSession EndpointSession {get; set;}
```

However, acc to http://docs.particular.net/nservicebus/messaging/send-a-message#outside-a-message-handler, IEndpointInstance is not registered. What should the user do? 

Can we create a sample using on of the containers to illustrate this scenario?
Because this is going to be a common scenario for a lot of our users who use the NServiceBus.Host.

**Comment by indualagarsamy** [comment]
@Particular/nservicebus-maintainers - please comment.

**Comment by indualagarsamy** [comment]
cc / @SeanFeldman

**Comment by danielmarbach** [comment]
I think we should create small snippets showing how to bind the necessary communication infrastructure to the supported containers. The question is where would you as a user look for that kind of information? Would it be the upgrade guide or a section under the corresponding containers, or both, or even another place?

**Comment by SeanFeldman** [comment]
My vote goes for both.
While upgrade guide was skimmed through, it is an overwhelming amount of information. Instinctively I'd go to a specific container to find that info. Yes, it would be redundant, but removal of auto-injected `IBus` is not a small thing for the customers.

**Comment by indualagarsamy** [comment]
@Particular/nservicebus-maintainers - 

A lot of our users are going to have a common component library that has an injected IBus like this:
https://github.com/indualagarsamy/MigrateIBus/blob/master/MigrateIBus/SharedCommonUsingV5Core/SomeComponentThatUsesTheBus.cs#L6

That gets used in the host, for example:
https://github.com/indualagarsamy/MigrateIBus/blob/master/MigrateIBus/NsbHostUsingV5Core/Bootstrapper.cs#L16

Can you please add enough information on how that can be migrated over to V6?
What's the correct way?

Is floating the appropriate context in the interface methods the only right approach? Or is there any other way in which this can be injected.

**Comment by danielmarbach** [comment]
I was only a specialist in the handler context task-force. I had a meeting with @mat-mcloughlin and @hmemcpy today to get more context. As far as I remember from talking to @timbussmann the samples about injection got scoped out from the handler context. For some reasons it seems it fell between the cracks after the handler context story was finished.

At the time of the discussion our hope was that the floating state story and the handler context story should give enough information for people to know what to do. This is of course silly. @SimonCropp said it multiple times that a blog post is not documentation. Nonetheless here is the blog post I wrote

http://www.planetgeek.ch/2016/04/26/avoid-threadstatic-threadlocal-and-asynclocal-float-the-state-instead/

and the handler context story currently in editorial review

https://github.com/Particular/Stories/blob/handler-context/blog-posts/HandlerContext/HandlerContext.md

The TL;DR version is

* Register `IEndpointInstance` or `IMessageSession` on the container for endpoint operations outside the scope of a handler
* DO NOT try to magically bind the handler context to the container. If you need access to the handler context float the context into your call hierarchy over "method injection". We do not want to show snippets how to bind the handler context because it would encourage bad practices. Sending messages is a handler concern. Not something that should leak into the dependency hierarchies.
* Using  `IEndpointInstance` or `IMessageSession` inside the call hierarchy of a handler is equivalent to immediate dispatch! So if the handler rolls back the messages will already be sent out.

I already pinged you on slack about the call.

**Comment by indualagarsamy** [comment]
Blocked by https://github.com/Particular/NServiceBus.Host/issues/85

**Comment by indualagarsamy** [comment]
Closing this issue based on @andreasohlund's comment here: https://github.com/Particular/NServiceBus.Host/issues/85#issuecomment-228681740
